axlearn/common/launch_trainer_main.py

```diff
@@ -4,18 +4,21 @@
 
 from absl import app, flags
 
-from axlearn.common import launch, launch_trainer, measurement
+from axlearn.common import launch, launch_trainer, measurement, monitoring
 from axlearn.common.config import config_for_function
 
 
 def main(_):
     measurement.initialize(flags.FLAGS)
+    monitoring.initialize(flags.FLAGS)
```

Comment:
Do we need a separate module for this, or can we consolidate into the `measurement` interface?

Reply:
I think it's a design choice. `measurement` part calls the GoodPut Recorder and uploads timestamps to Cloud Logging. `monitoring` part calls the GoodPut Monitor, which calculates goodput for a customized time interval and regularly updates Tensorboard file in GCS. These seem to have separate functionalities and sometimes users might need only `measurement` and not `monitoring`

Reply:
Can the decision of whether to upload be handled via a config field instead of a separate module?

Reply:
The config is useful either way to decide whether/how to upload the data. The intention to separate the monitoring module is to allow for the flexibility of using other types of monitors and allow for growing functionality of monitoring which is likely to depend on different things than recording. I think its useful to be able to configure these two entities separately because of the dataflow and the underlying systems that support the recording vs the calculating/monitoring, adding future flexibility and overall modularity. Having said that, its definitely possible to couple them in the same module, but I am not sure if its a better overall design. Let me know your thoughts though! :)

Reply:
Thanks for the discussion!

> allow for the flexibility of using other types of monitors and allow for growing functionality of monitoring which is likely to depend on different things than recording

Our config system supports composition of different configurable objects. For example, you can imagine that in the measurement module, we have a configurable object that is composed of a recorder and a monitor. A simple design might look like:

```
class GoodputMeasurement(...):
    @config_class
    class Config(...):
        recorder: Recorder.Config  # Configures recorder.
        monitor: Monitor.Config  # Configures monitor.
```

This allows us to interchange monitor implementations while keeping the same recorder, or vice versa, without introducing a separate Python module. (As it is, it seems tolerable to further simplify by extending `GoodputRecorder` with the `monitor` config directly, which may avoid additional pass-through APIs for `record` and `start_monitoring`. If the `Recorder` name is confusing, we can rename it.)

That said, I wonder what other monitors we plan to support for goodput. Are there alternatives to the `goodput_monitoring` that we plan to add that demands flag-level configurability? Or can we start with the simpler option of using the `goodput_monitoring` as the default option for now?

> because of the dataflow and the underlying systems that support the recording vs the calculating/monitoring

As it is currently implemented, we seem to want to instantiate and run both at the same time anyway. We can always add another Python module later if it grows too complex.

WDYT?

Reply:
Thanks! I agree and apologize for the delay in response, I will make the changes accordingly!